Pull Request for the Structs for the Generation Algorithm

edward3-uchi opened the pull request:
OK! I am finally done. This Pull Request contains the structs for the generation algorithm for the rpg/openworld team. The pull request affects only two different directories: openworld and the root CMakeLists file. The contexts within this pull request are the following: 

gencontext:

this struct contains the information required for the algorithm to autogenerate a room. It contains the player's level, the linked list of roomspecs (which I will explain next), number of openpaths to be made in the room, the path we are linking this room to, and the number of npcs to generate in the room. 

roomspec: 

this struct contains information for generating a room. all of the roomspecs will be hardcoded, meaning this data will live somewhere (TDB) in the openworld module. The roomspecs contain the hashtables for path and items, a long description, and a short description.

speclist:

this struct is a linked list of roomspecs. in the future, all of the speclists will be organized by some certain theme. for example, all of the roomspecs in one speclist may contain rooms that are fit for a dungeon theme or castle theme. the linked list allows the roomspecs to be listed in a manner that we can pull one at random and conserves memory.

CCalderon01 commented:
Looks good! The only feedback I have is for gen_structs.h:
- I'm apprehensive of your roomspec struct. I think it's very repetitive of the room struct that you can find not only in the sample_rooms.h file but also in /game-state/room.h. If this is just a temp fix until the room module is complete for the openworld feature, I understand, but even then I think you could have just referenced the struct in game-state/room.h.
- Also, just for continuity, you might want to fix lines 43-47  in gen_structs.h so that you define speclist in the same way that you define every other struct:
ie.e. typedef struct speclist{ ....} speclist_t;
(instead of inserting an additional line just to name the type)

CCalderon01 commented on include/openworld/gen_structs.h:
```diff
@@ -0,0 +1,200 @@
+ * - roomspec_t *spec: pointer to some room specification
+ * - speclist_t *next: pointer to the next part of the list.
+ */
+typedef struct speclist speclist_t;
```
Just for consistensy, you might want to fix lines 43-47 in gen_structs.h so that you define speclist in the same way that you define every other struct:
ie.e. typedef struct speclist{ ....} speclist_t;
(instead of inserting an additional line just to name the type)

edward3-uchi replied:
Ah, okay maybe I can implement it a different way. I was worried that I had to implement it this way since a speclist_t contained another speclist_t. Like it was the method for getting around compile-time errors with recursive structs.

dwahme replied:
Gonna hop in here- the definition would look something like
```
typedef struct speclist {
    roomspec_t *spec;
    struct speclist *next;
    struct speclist *prev;
} speclist_t;
```

dwahme replied:
(always use doubly linked lists over singly linked ones if you're getting them for free)

edward3-uchi replied:
Ok great gonna make these changes and commit them. I saw that these changes were just made earlier when I looked at Brians's message. I didn't even think to use double linked lists that seems so much better.

dwahme commented on include/openworld/gen_structs.h:
```diff
@@ -0,0 +1,200 @@
+ * See chiventure/src/openworld/gen_structs.c source code to see implementation.
+ */
+
+#include "../game-state/game_state_common.h"
+#include "../game-state/game.h"
```
These should just be `#include "game-state/game_state_common.h"`

dwahme commented on include/openworld/gen_structs.h:
```diff
@@ -0,0 +1,200 @@
+
+#include "../game-state/game_state_common.h"
+#include "../game-state/game.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
```
For `#includes`, standard libraries come first, then user-defined ones. Also, they should be in alphabetical order
```
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include "game-state/game.h"
#include "game-state/game_state_common.h"
```

dwahme commented on include/openworld/gen_structs.h:
```diff
@@ -0,0 +1,200 @@
+ * - item_hash_t *items: hash table of items in room
+ * - path_hash_t *paths: hash table of paths in room
+ */
+typedef struct roomspec {
+    char *short_desc;
+    char *long_desc;
+    item_hash_t *items;
+    path_hash_t *paths;
+} roomspec_t;
```
I don't think this needs a `path_hash_t`- paths will be added to the room that's generated from the roomspec when it's actually hooked into the game

dwahme commented on include/openworld/gen_structs.h:
```diff
@@ -0,0 +1,200 @@
+    path_t *path;
+    int level;
+    int openpaths;
+    int numnpcs;
```
This is something that should live in the roomspec, not the gencontext

edward3-uchi replied:
Ah, okay that makes more sense. I was envisioning these variables as being ways for the autogenerate to fine-tune the selection from speclist_t, but I can see why it would make more sense for variables like these to live in the roomspec. However, I'm a little confused as to why we would want to put the path variable in the roomspec. The path is the path we are attaching our room to. Should that live in roomspec as well? That value isn't necessarily going to be hardcoded anywhere.

dwahme commented on src/openworld/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+#add_library(openworld
+#            src/autogenerate.c
+#            ../../../tests/openworld/test_autogenerate.c)
```
Don't have commented-out code in a PR

dwahme commented on src/openworld/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+target_link_libraries(openworld game-state)
+target_link_libraries(openworld common)
+
+# TODO: Checkpointing (which CLI depends on) uses a private header file
+#       from game-state. See TODO in checkpointing CMakeLists.txt
+#target_include_directories(cli PRIVATE ../game-state/src/)
```
Don't have TODOs in a PR

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+/* see gen_structs.h */
+gencontext_t* gencontext_new(path_t *path, int level, int openpaths, int numnpcs, speclist_t *speclist){
+
+    gencontext_t *contextnew = calloc(1, sizeof(gencontext_t));
```
You need to check to make sure that `calloc` didn't return NULL

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+gencontext_t* gencontext_new(path_t *path, int level, int openpaths, int numnpcs, speclist_t *speclist){
+
+    gencontext_t *contextnew = calloc(1, sizeof(gencontext_t));
+    contextnew->path = path;
+    contextnew->level = level;
+    contextnew->openpaths = openpaths;
+    contextnew->numnpcs = numnpcs;
+    contextnew->speclist = speclist;
```
This should call `gencontext_init()`

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+/* see gen_structs.h */
+roomspec_t* roomspec_new(char *short_desc, char *long_desc, item_hash_t *items, path_hash_t *paths){
+
+    roomspec_t *roomspecnew = calloc(1, sizeof(roomspec_t));
```
You need to check that `calloc` didn't return `NULL`

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+roomspec_t* roomspec_new(char *short_desc, char *long_desc, item_hash_t *items, path_hash_t *paths){
+
+    roomspec_t *roomspecnew = calloc(1, sizeof(roomspec_t));
+    roomspecnew->short_desc = short_desc;
+    roomspecnew->long_desc = long_desc;
+    roomspecnew->items = items;
+    roomspecnew->paths = paths;
```
This should be a call to `roomspec_init`

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+/* see gen_structs.h */
+speclist_t* speclist_new(roomspec_t *spec){
+
+    speclist_t *listnew = calloc(1, sizeof(speclist_t));
```
Check for `calloc()` returning `NULL`

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+speclist_t* speclist_new(roomspec_t *spec){
+
+    speclist_t *listnew = calloc(1, sizeof(speclist_t));
+    listnew->spec = spec;
+    listnew->prev = NULL;
+    listnew->next = NULL;
```
Call `speclist_init()`

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+
+    free(list);
+    return SUCCESS;	
+}
```
You should provide a `speclist_free_all()` that takes in a linked list of `speclist`'s and frees all of them. Look into `DL_FOREACH_SAFE` and `DL_DELETE`

dwahme commented on tests/openworld/test_structs.c:
```diff
@@ -0,0 +1,128 @@
+
+    cr_assert_not_null(path, "path was NULL\n");
+
+    gencontext_t *context = malloc(sizeof(gencontext_t));
```
`calloc`

edward3-uchi replied:
not sure why the test is showing this file... the newest commit had changed these to calloc. Hopefully this commit will update these correctly.

dwahme commented on tests/openworld/test_structs.c:
```diff
@@ -0,0 +1,128 @@
+ * can be initialized successfully. */
+Test(roomspec, init1){
+
+    roomspec_t *spec = malloc(sizeof(roomspec_t));
```
`calloc`

dwahme commented on tests/openworld/test_structs.c:
```diff
@@ -0,0 +1,128 @@
+
+    cr_assert_not_null(spec, "failed to create new roomspec_t\n");
+
+    speclist_t *list = malloc(sizeof(speclist_t));
```
`calloc`

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+    if (spec == NULL)
+	return FAILURE;
+
+    spec->short_desc = short_desc;
```
Generally, it's not a great idea to trust strings given to you as a function argument. Instead, you should allocate space for the string, and then copy it in so that the user is free to do whatever they want to the string afterwards. It should look something like this

```
spec->short_desc = calloc(MAX_SHORT_DESC_LEN + 1, sizeof(char));
if (spec->short_desc == NULL)
{
    roomspec_free(spec);
    return FAILURE;
}
strncpy(spec->short_desc, short_desc, MAX_SHORT_DESC_LEN);
```

edward3-uchi replied:
Oh, wow. I never knew this. Thanks! That definitely seems like a good idea...

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+	return FAILURE;
+
+    spec->short_desc = short_desc;
+    spec->long_desc = long_desc;
```
Deep copy the string

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,120 @@
+
+    if (spec == NULL)
+        return FAILURE;
+
```
This will need to free `spec->short_desc` and `spec->long_desc` if they're being deep copied. However, they may only be partially initialized (think like `calloc` fails in `roomspec_init`. So, it'll need to be something like
```
if (spec->short_desc)
{
    free(spec->short_desc);
}
```

edward3-uchi commented:
I just made a commit that addressed all of the recommended changes. There are updates to the structs, added a new funcion that frees the entire speclist, and made a new test for it as well. Cleaned up the CMake in src and some other cleaning issues as well.

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,168 @@
+    speclist_t *tmp, *ptmp;
+    int check = 0;
+
+    for(tmp = list; tmp != NULL; tmp = tmp->next){
+	ptmp = tmp;	
+        check = speclist_free(ptmp);
+        if (check == FAILURE){
+   	    return check;
+        }
+    }
```
This should use `DL_FOREACH_SAFE` and `DL_DELETE` (provided by utlist- https://troydhanson.github.io/uthash/utlist.html)

Basically, it should look something like
```
speclist *elt, *tmp;
DL_FOREACH_SAFE(list, elt, tmp) {
    DL_DELETE(head,elt);
    speclist_free(elt);
}
```

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,168 @@
+
+    spec->short_desc = calloc(MAX_SDESC_LEN + 1, sizeof(char));
+    if (spec->short_desc == NULL){
+	roomspec_free(spec);
+	fprintf(stderr, "calloc failed to allocate space for spec's short_desc. \n");
+	return FAILURE;
```
No tab characters- this should be 8 spaces

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,168 @@
+int init_speclist(speclist_t *list, roomspec_t *spec){
+
+    if (list == NULL)
+	return FAILURE;
```
No tab characters

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,168 @@
+int gencontext_free(gencontext_t *context){
+
+    if (context == NULL)
+	return FAILURE;
```
No tab characters- this should be 8 spaces

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,168 @@
+        return FAILURE;
+
+    if (spec->short_desc){
+	free(spec->short_desc);
```
No tab characters

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,168 @@
+int init_roomspec(roomspec_t *spec, char *short_desc, char *long_desc, item_hash_t *items){
+    
+    if (spec == NULL)
+	return FAILURE;
```
No tab characters

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,168 @@
+    speclist_t *listnew = calloc(1, sizeof(speclist_t));
+    
+    if (listnew == NULL){
+	fprintf(stderr, "calloc failed to allocate space for listnew. \n");
```
No tab characters

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,168 @@
+int speclist_free(speclist_t *list){
+    
+    if (list == NULL)
+	return FAILURE;
```
No tab characters

dwahme commented on src/openworld/src/gen_structs.c:
```diff
@@ -0,0 +1,168 @@
+int speclist_free_all(speclist_t *list){
+
+    if (list == NULL)
+	return FAILURE;
```
No tab characters

dwahme commented on tests/openworld/test_structs.c:
```diff
@@ -0,0 +1,163 @@
+    gencontext_t *context = calloc(1, sizeof(gencontext_t));
+    
+    if (context == NULL){
+	fprintf(stderr, "failed to calloc for context. \n");
```
No tab characters

dwahme commented on tests/openworld/test_structs.c:
```diff
@@ -0,0 +1,163 @@
+    roomspec_t *spec = calloc(1, sizeof(roomspec_t));
+   
+    if (spec == NULL){
+	fprintf(stderr, "failed to calloc for spec. \n");
```
No tab characters

dwahme commented:
Issue Score: ✔️+

Comments:
Great job! Here are a few things you may want to improve in future issues:

When an issue is done, you should include a final comment summarizing what was done, and why the issue is being closed.

<!-- {"score": "check-plus"} -->
